File: November LeetCoding Challenge(2020)/Day 19 Decode String.java
class Solution {
    public boolean nums(char x){
        return x >= '0' && x <= '9';
    }
    public String decodeString(String s) {
        Stack<Integer> countStack = new Stack<>();
        Stack<StringBuilder> strStack = new Stack<>();
        StringBuilder cur = new StringBuilder();
        int k = 0;
        for (char ch : s.toCharArray()) {
            if (Character.isDigit(ch)) {
                k = k * 10 + ch - '0';
            } else if ( ch == '[') {
                countStack.push(k);
                strStack.push(cur);
                cur = new StringBuilder();
                k = 0;
            } else if (ch == ']') {
                StringBuilder tmp = cur;
                cur = strStack.pop();
                for (k = countStack.pop(); k > 0; --k) cur.append(tmp);
            } else cur.append(ch);
        }
        return cur.toString();
    }
}

File: December LeetCoding Challenge(2021)/Day 19 Decode String.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<int> numStack;
        stack<string> strStack;
        
        int k = 0;
        string cur = "";
        for(auto ch : s){
            if (isdigit(ch)){
                k = k * 10 + (ch - '0');
            }else if (ch == '['){
                numStack.push(k);
                k = 0;
                strStack.push(cur);
                cur = "";
            }else if (ch == ']'){
                string temp = cur;
                cur = strStack.top(); strStack.pop();
                for(int i=numStack.top(); i>0; i--){
                    cur += temp;
                }
                numStack.pop();
            }else{
                cur += ch;
            }
        }
        return cur;
    }
};
